File: ts-reference/tests/example-program-gen/exp/instructions/initializeWithValues.ts
import { TransactionInstruction, PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface InitializeWithValuesArgs {
  boolField: boolean
  u8Field: number
  i8Field: number
  u16Field: number
  i16Field: number
  u32Field: number
  i32Field: number
  f32Field: number
  u64Field: BN
  i64Field: BN
  f64Field: number
  u128Field: BN
  i128Field: BN
  bytesField: Array<number>
  stringField: string
  pubkeyField: PublicKey
  vecField: Array<BN>
  vecStructField: Array<types.FooStructFields>
  optionField: boolean | null
  optionStructField: types.FooStructFields | null
  structField: types.FooStructFields
  arrayField: Array<boolean>
  enumField1: types.FooEnumKind
  enumField2: types.FooEnumKind
  enumField3: types.FooEnumKind
  enumField4: types.FooEnumKind
}

export interface InitializeWithValuesAccounts {
  state: PublicKey
  nested: {
    clock: PublicKey
    rent: PublicKey
  }
  payer: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.bool("boolField"),
  borsh.u8("u8Field"),
  borsh.i8("i8Field"),
  borsh.u16("u16Field"),
  borsh.i16("i16Field"),
  borsh.u32("u32Field"),
  borsh.i32("i32Field"),
  borsh.f32("f32Field"),
  borsh.u64("u64Field"),
  borsh.i64("i64Field"),
  borsh.f64("f64Field"),
  borsh.u128("u128Field"),
  borsh.i128("i128Field"),
  borsh.vecU8("bytesField"),
  borsh.str("stringField"),
  borsh.publicKey("pubkeyField"),
  borsh.vec(borsh.u64(), "vecField"),
  borsh.vec(types.FooStruct.layout(), "vecStructField"),
  borsh.option(borsh.bool(), "optionField"),
  borsh.option(types.FooStruct.layout(), "optionStructField"),
  types.FooStruct.layout("structField"),
  borsh.array(borsh.bool(), 3, "arrayField"),
  types.FooEnum.layout("enumField1"),
  types.FooEnum.layout("enumField2"),
  types.FooEnum.layout("enumField3"),
  types.FooEnum.layout("enumField4"),
])

export function initializeWithValues(
  args: InitializeWithValuesArgs,
  accounts: InitializeWithValuesAccounts
) {
  const keys = [
    { pubkey: accounts.state, isSigner: true, isWritable: true },
    { pubkey: accounts.nested.clock, isSigner: false, isWritable: false },
    { pubkey: accounts.nested.rent, isSigner: false, isWritable: false },
    { pubkey: accounts.payer, isSigner: true, isWritable: true },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([220, 73, 8, 213, 178, 69, 181, 141])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      boolField: args.boolField,
      u8Field: args.u8Field,
      i8Field: args.i8Field,
      u16Field: args.u16Field,
      i16Field: args.i16Field,
      u32Field: args.u32Field,
      i32Field: args.i32Field,
      f32Field: args.f32Field,
      u64Field: args.u64Field,
      i64Field: args.i64Field,
      f64Field: args.f64Field,
      u128Field: args.u128Field,
      i128Field: args.i128Field,
      bytesField: Buffer.from(args.bytesField),
      stringField: args.stringField,
      pubkeyField: args.pubkeyField,
      vecField: args.vecField,
      vecStructField: args.vecStructField.map((item) =>
        types.FooStruct.toEncodable(item)
      ),
      optionField: args.optionField,
      optionStructField:
        (args.optionStructField &&
          types.FooStruct.toEncodable(args.optionStructField)) ||
        null,
      structField: types.FooStruct.toEncodable(args.structField),
      arrayField: args.arrayField,
      enumField1: args.enumField1.toEncodable(),
      enumField2: args.enumField2.toEncodable(),
      enumField3: args.enumField3.toEncodable(),
      enumField4: args.enumField4.toEncodable(),
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: tests/client_gen/example_program_gen/instructions/initialize_with_values.py
from __future__ import annotations
import typing
from solders.pubkey import Pubkey
from solders.system_program import ID as SYS_PROGRAM_ID
from solders.sysvar import RENT, CLOCK
from solders.instruction import Instruction, AccountMeta
from anchorpy.borsh_extension import BorshPubkey
from construct import Construct
import borsh_construct as borsh
from .. import types
from ..program_id import PROGRAM_ID


class InitializeWithValuesArgs(typing.TypedDict):
    bool_field: bool
    u8_field: int
    i8_field: int
    u16_field: int
    i16_field: int
    u32_field: int
    i32_field: int
    f32_field: float
    u64_field: int
    i64_field: int
    f64_field: float
    u128_field: int
    i128_field: int
    bytes_field: bytes
    string_field: str
    pubkey_field: Pubkey
    vec_field: list[int]
    vec_struct_field: list[types.foo_struct.FooStruct]
    option_field: typing.Optional[bool]
    option_struct_field: typing.Optional[types.foo_struct.FooStruct]
    struct_field: types.foo_struct.FooStruct
    array_field: list[bool]
    enum_field1: types.foo_enum.FooEnumKind
    enum_field2: types.foo_enum.FooEnumKind
    enum_field3: types.foo_enum.FooEnumKind
    enum_field4: types.foo_enum.FooEnumKind


layout = borsh.CStruct(
    "bool_field" / borsh.Bool,
    "u8_field" / borsh.U8,
    "i8_field" / borsh.I8,
    "u16_field" / borsh.U16,
    "i16_field" / borsh.I16,
    "u32_field" / borsh.U32,
    "i32_field" / borsh.I32,
    "f32_field" / borsh.F32,
    "u64_field" / borsh.U64,
    "i64_field" / borsh.I64,
    "f64_field" / borsh.F64,
    "u128_field" / borsh.U128,
    "i128_field" / borsh.I128,
    "bytes_field" / borsh.Bytes,
    "string_field" / borsh.String,
    "pubkey_field" / BorshPubkey,
    "vec_field" / borsh.Vec(typing.cast(Construct, borsh.U64)),
    "vec_struct_field"
    / borsh.Vec(typing.cast(Construct, types.foo_struct.FooStruct.layout)),
    "option_field" / borsh.Option(borsh.Bool),
    "option_struct_field" / borsh.Option(types.foo_struct.FooStruct.layout),
    "struct_field" / types.foo_struct.FooStruct.layout,
    "array_field" / borsh.Bool[3],
    "enum_field1" / types.foo_enum.layout,
    "enum_field2" / types.foo_enum.layout,
    "enum_field3" / types.foo_enum.layout,
    "enum_field4" / types.foo_enum.layout,
)


class InitializeWithValuesAccounts(typing.TypedDict):
    state: Pubkey
    payer: Pubkey


def initialize_with_values(
    args: InitializeWithValuesArgs,
    accounts: InitializeWithValuesAccounts,
    program_id: Pubkey = PROGRAM_ID,
    remaining_accounts: typing.Optional[typing.List[AccountMeta]] = None,
) -> Instruction:
    keys: list[AccountMeta] = [
        AccountMeta(pubkey=accounts["state"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=CLOCK, is_signer=False, is_writable=False),
        AccountMeta(pubkey=RENT, is_signer=False, is_writable=False),
        AccountMeta(pubkey=accounts["payer"], is_signer=True, is_writable=True),
        AccountMeta(pubkey=SYS_PROGRAM_ID, is_signer=False, is_writable=False),
    ]
    if remaining_accounts is not None:
        keys += remaining_accounts
    identifier = b"\xdcI\x08\xd5\xb2E\xb5\x8d"
    encoded_args = layout.build(
        {
            "bool_field": args["bool_field"],
            "u8_field": args["u8_field"],
            "i8_field": args["i8_field"],
            "u16_field": args["u16_field"],
            "i16_field": args["i16_field"],
            "u32_field": args["u32_field"],
            "i32_field": args["i32_field"],
            "f32_field": args["f32_field"],
            "u64_field": args["u64_field"],
            "i64_field": args["i64_field"],
            "f64_field": args["f64_field"],
            "u128_field": args["u128_field"],
            "i128_field": args["i128_field"],
            "bytes_field": args["bytes_field"],
            "string_field": args["string_field"],
            "pubkey_field": args["pubkey_field"],
            "vec_field": args["vec_field"],
            "vec_struct_field": list(
                map(lambda item: item.to_encodable(), args["vec_struct_field"])
            ),
            "option_field": args["option_field"],
            "option_struct_field": (
                None
                if args["option_struct_field"] is None
                else args["option_struct_field"].to_encodable()
            ),
            "struct_field": args["struct_field"].to_encodable(),
            "array_field": args["array_field"],
            "enum_field1": args["enum_field1"].to_encodable(),
            "enum_field2": args["enum_field2"].to_encodable(),
            "enum_field3": args["enum_field3"].to_encodable(),
            "enum_field4": args["enum_field4"].to_encodable(),
        }
    )
    data = identifier + encoded_args
    return Instruction(program_id, data, keys)
